library/common/main_interface.h

```diff
@@ -1,12 +1,214 @@
 #pragma once
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
 
 // NOLINT(namespace-envoy)
 
+/**
+ * Handle to an Envoy engine instance. Valid only for the lifetime of the engine and not intended
+ * for any external interpretation or use.
+ */
+typedef uint64_t envoy_engine_t;
+
+/**
+ * Handle to an outstanding Envoy HTTP stream. Valid only for the duration of the stream and not
+ * intended for any external interpretation or use.
+ */
+typedef uint64_t envoy_stream_t;
+
+/**
+ * Result codes returned by all calls made to this interface.
+ */
+typedef enum { ENVOY_SUCCESS, ENVOY_FAILURE } envoy_status_t;
+
+/**
+ * Error code associated with terminal status of a HTTP stream or request.
+ */
+typedef enum {} envoy_error_code_t;
+
+/**
+ * Common abstraction for strings.
+ */
+typedef struct {
+  uint64_t length;
+  char* data;
+} envoy_string;
+
+/**
+ * Holds data about an HTTP request.
```

Comment:
Similarly throughout this file, if types are specific to HTTP do you think that detail should be reflected in their name? Or maybe omitted from the docstrings?

Reply:
The reason why I have not done this is because I feel including http in the type name might be misleading some places. For instance in the error code type we are could create misunderstanding between it and actual http codes (5xx, etc).